2021/day16/day16.java

```diff
@@ -0,0 +1,151 @@
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Scanner;
+
+public class day16 {
+    public static class Packet {
+        Long value;
+        int length;
+        int versionSum;
+        String binData;
+    }
+
+    public static String readInput(String filename) {
+        String s = "";
+        String result = "";
+        Scanner sc = null;
+        try {
+            sc = new Scanner(new File(filename));
+            s = sc.next();
+            for (char c : s.toCharArray()) {
+                result += hexToBin.get(String.valueOf(c));
+            }
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        } finally {
+            sc.close();
+        }
+        return result;
+    }
+
+    private static final HashMap<String, String> hexToBin = new HashMap<>();
+    static {
+        hexToBin.put("0", "0000");
+        hexToBin.put("1", "0001");
+        hexToBin.put("2", "0010");
+        hexToBin.put("3", "0011");
+        hexToBin.put("4", "0100");
+        hexToBin.put("5", "0101");
+        hexToBin.put("6", "0110");
+        hexToBin.put("7", "0111");
+        hexToBin.put("8", "1000");
+        hexToBin.put("9", "1001");
+        hexToBin.put("A", "1010");
+        hexToBin.put("B", "1011");
+        hexToBin.put("C", "1100");
+        hexToBin.put("D", "1101");
+        hexToBin.put("E", "1110");
+        hexToBin.put("F", "1111");
+    }
+
+    public static Long makeCalculations(int packetType, ArrayList<Long> values) {
+        switch (packetType) {
+            case 0:
+                Long sum = 0L;
+                for (Long num : values) {
+                    sum += num;
+                }
+                return sum;
+            case 1:
+                Long product = 1L;
+                for (Long num : values) {
+                    product *= num;
+                }
+                return product;
+            case 2:
+                return Collections.min(values);
+            case 3:
+                return Collections.max(values);
+            case 5:
+                return values.get(0) > values.get(1) ? 1L : 0L;
+            case 6:
+                return values.get(0) < values.get(1) ? 1L : 0L;
+            case 7:
+                return values.get(0).equals(values.get(1)) ? 1L : 0L;
+            default:
+                System.out.println("Error in switch statement");
+                return -1L;
+        }
+    }
```

Comment:
This begs for a stream in cases 0-3:

```java
LongStream valueStream = values.stream();
switch (packetType) {
  case 0 -> valueStream.sum();
  case 1 -> valueStream.reduce(1, (v, acc) -> v * acc));
  case 2 -> valueStream.min().orElseThrow();
  case 3 -> valueStream.max().orElseThrow();
}
```

but I understand that's not the focus right now. Nonetheless, make sure to check out enhanced switch, looks nicer to me

Reply:
Enhanced switch was introduced in Java 13 or 14 and I am pretty sure the defautl one in Ubuntu is 11, so no enhanced switch for me.